### ISSUE TITLE
Store PackedByteArray array in more optimal format for tscn files (Ex: Base64)

### ISSUE THREAD

**Issue by jitspoe** [godotengine/godot-proposals#8053]
### Describe the project you are working on

Retro FPS and 2.5D platformer.

### Describe the problem or limitation you are having in your project

When using a tool to generate geometry, the data typically gets stored in the tscn file as a packed byte array, which currently is comma separated list.  Ex:
```
"vertex_data": PackedByteArray(0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 255, 127, 255, 255, 255, 191, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 255, 255, 255, 255, 255, 255, 63, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 255, 255, 127, 255, 255, 255, 255, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 0, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 12, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 12, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 0, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 12, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 0, 194, 0, 0, 255, 127, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 0, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 170, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 166, 65, 0, 0, 0, 0, 0, 0, 12, 194, 255, 127, 0, 0, 255, 127, 255, 191, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 255, 255, 255, 255, 255, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 255, 255, 255, 255, 255, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 127, 255, 255, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 0, 194, 255, 127, 255, 255, 255, 255, 255, 255, 0, 0, 166, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 127, 255, 255, 255, 255, 255, 255, 0, 0, 170, 65, 0, 0, 128, 64, 0, 0, 12, 194, 255, 127, 255, 255, 255, 255, 255, 255)
}]
```
This can severely bloat the file sizes.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

By using something like Base64, or even just 2-byte hex values, to store the data, the file size would be considerably reduced (and likely speed up save and load times, too).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

From the end user perspective, nothing would change.  The tscn files would just be written with something like:
"PackedByteArrayB64(wHateVer..." instead of "PackedByteArray(1, 2, 3, 4, ..."
Obviously the old format would need to be supported for backward compatibility, which is why I suggested a different name, like "PackedByteArrayB64"

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Core feature.  Can't be easily worked around for storing native types like meshes.

### Is there a reason why this should be core and not an add-on in the asset library?

More optimal files should just be the default. :)

**Comment by Mickeon** [comment]
What could be done, perhaps, is storing the **PackedByteArray** in a more compact format after a certain amount of elements. Assuming the definition would use a different, I feel like it can be an arbitrary amount just fine.

It's worth noting `.tscn` files are designed to be editable in any text editor of choice, too. Doing this may affect this purpose, hence the above suggestion.

Other Resources/Nodes that could benefit from this include Image, FontFile ([partially addresses this](https://github.com/godotengine/godot/issues/82254)), TileMap...

**Comment by KoBeWi** [comment]
A similar idea was discussed on the chat:
https://chat.godotengine.org/channel/core?msg=q4HszqDNFsaJgGSEN
tl;dr is that if you have a binary data, you should store it in a separate binary file, not inside scene.

**Comment by Mickeon** [comment]
The problem is, this cannot be always be accomplished because of Godot's internal meddling. In the case of TileMap, a similar problem occurs, even if it's not technically using a byte array. There's many tiles, each composed of multiple values. It does bloat up the filesize of the `.tres` file by a large margin, and it can be argued that modifying these numbers with a text editor is just as "meaningless" as if they were in a condensed format.

**Comment by jitspoe** [comment]
> A similar idea was discussed on the chat: https://chat.godotengine.org/channel/core?msg=q4HszqDNFsaJgGSEN tl;dr is that if you have a binary data, you should store it in a separate binary file, not inside text scene.

I don't think this brought me to the appropriate message.  Seems to just be the most recent chat in #core, which was about structs in gdscript.

For context, most of the stuff in my tscn is just placed meshes and other scenes, but I also have a vine generator where I can click on a couple points and it will generate a vine mesh between them. This needs to be done in the context of the scene with all the other meshes present, so I can't really create them as a separate scn file, and I want to be able to edit and move things around, so creating the vines in the scene, then saving them as a separate scene isn't an ideal workflow if I want to make adjustments in the future.  If I mark them as editable and edit them in the scene, we're back to the first problem of the PackedByteArray for the mesh data stored a comma separated list in the tscn file.

**Comment by KoBeWi** [comment]
> I don't think this brought me to the appropriate message. Seems to just be the most recent chat in #core, which was about structs in gdscript.

Yeah the link is a bit unreliable. You need to wait a bit until it loads and try few times.

**Comment by Calinou** [comment]
I recall Base64 serialization being used a long time ago in Godot (maybe in Godot 2.x in `.xscn` and `.tscn`), but I'm not 100% sure.

**Comment by Mickeon** [comment]
This proposal has been implemented by https://github.com/godotengine/godot/pull/89186 and should be safe to close.
